### PR TITLE
Fix sara r4 error msg

### PR DESCRIPTION
**Pull request by chinglee-iot**
* **Filter empty line in pktio**
SARA-R4 sends "\0\0" after reboot. These will be filter out in pktio and will not be regarded as AT_UNDEFINED message.
* **Add CELLULAR_CONFIG_SARA_R4_REBOOT_ON_INIT**
Some platforms may not support hard reboot on SARA-R4. This config helps to trigger a soft reboot during module init.
*  **Print warning message for socket not closed properly in previous program execution**
If the socket is not closed properly in previous program execution, socket close or socket data URC will be received in new program execution. Print warning message to indicate socket not closed properly.

### CODE REVIEW

**Review by ActoryOu — APPROVED** [review]

File: modules/sara_r4/cellular_r4.c [hunk 1]
```diff
@@ -314,6 +314,10 @@ CellularError_t rebootCellularModem( CellularContext_t * pContext,
 
             break;
         }
+        else
+        {
+            LogInfo( ( "rebootCellularModem: Modem is not ready. Retry sending ATE0." ) );
```
Comment (aggarg) [inline modules/sara_r4/cellular_r4.c:319]: Should it be a LogWarn?

File: modules/sara_r4/cellular_r4_urc_handler.c [hunk 2]
```diff
@@ -505,22 +505,30 @@ static void _cellular_UrcProcessUusord( CellularContext_t * pContext,
         /* Call the callback function of this session. */
         if( atCoreStatus == CELLULAR_AT_SUCCESS )
         {
-            pSocketData = _Cellular_GetSocketData( pContext, socketIndex );
-
-            if( pSocketData == NULL )
+            if( socketIndex == INVALID_SOCKET_INDEX )
             {
-                LogError( ( "_cellular_UrcProcessUusord : invalid socket index %d", socketIndex ) );
+                LogWarn( ( "_cellular_UrcProcessUusord : unknow session data received. "
```
Comment (aggarg) [inline modules/sara_r4/cellular_r4_urc_handler.c:510]:
```suggestion
                LogWarn( ( "_cellular_UrcProcessUusord : unknown session data received. "
```

File: modules/sara_r4/cellular_r4_urc_handler.c [hunk 3]
```diff
@@ -567,22 +575,30 @@ static void _cellular_UrcProcessUusocl( CellularContext_t * pContext,
         /* Call the callback function of this session. */
         if( atCoreStatus == CELLULAR_AT_SUCCESS )
         {
-            pSocketData = _Cellular_GetSocketData( pContext, socketIndex );
-
-            if( pSocketData == NULL )
+            if( socketIndex == INVALID_SOCKET_INDEX )
             {
-                LogError( ( "_cellular_UrcProcessUusocl : invalid socket index %d", socketIndex ) );
+                LogWarn( ( "_cellular_UrcProcessUusocl : unknow session closed URC received. "
```
Comment (aggarg) [inline modules/sara_r4/cellular_r4_urc_handler.c:580]:
```suggestion
                LogWarn( ( "_cellular_UrcProcessUusocl : unknown session closed URC received. "
```

File: modules/sara_r4/cellular_r4_urc_handler.c [hunk 4]
```diff
@@ -505,22 +505,30 @@ static void _cellular_UrcProcessUusord( CellularContext_t * pContext,
...
                 {
-                    LogDebug( ( "_cellular_UrcProcessUusord: Data ready callback not set!!" ) );
+                    /* Indicate the upper layer about the data reception. */
+                    if( pSocketData->dataReadyCallback != NULL )
+                    {
+                        pSocketData->dataReadyCallback( pSocketData, pSocketData->pDataReadyCallbackContext );
+                    }
+                    else
+                    {
+                        LogDebug( ( "_cellular_UrcProcessUusord: Data ready callback not set!!" ) );
```
Comment (aggarg) [inline modules/sara_r4/cellular_r4_urc_handler.c:530]: Can it happen in a legitimate case? If not, it should probably be a warning or error. 

Can the application still function? If yes, it should be warning, else error.

**Review by ActoryOu — APPROVED** [review]

File: modules/sara_r4/cellular_r4_urc_handler.c (hunk 4 above)
Comment (chinglee-iot) [reply modules/sara_r4/cellular_r4_urc_handler.c:530]: The application may use polling method to read socket data. It is legal not to register the callback function.

File: modules/sara_r4/cellular_r4.c (hunk 1 above)
Comment (chinglee-iot) [reply modules/sara_r4/cellular_r4.c:319]: Update in new commit.

File: modules/sara_r4/cellular_r4_urc_handler.c (hunk 2 above)
Comment (chinglee-iot) [reply modules/sara_r4/cellular_r4_urc_handler.c:510]: Update in new commit.

File: modules/sara_r4/cellular_r4_urc_handler.c (hunk 3 above)
Comment (chinglee-iot) [reply modules/sara_r4/cellular_r4_urc_handler.c:580]: Update in new commit.

**Review by ActoryOu — DISMISSED** [review]

File: modules/sara_r4/cellular_r4_urc_handler.c (hunk 4 above)
Comment (aggarg) [reply modules/sara_r4/cellular_r4_urc_handler.c]: Apologies but then you were right and it should be debug.

**Review by aggarg — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Review by ActoryOu — APPROVED** [review]

**Pull request merged**